### PR TITLE
[Batch Viewer] Avoid multiple requests to Tenderly API

### LINKED ISSUE #121: Avoid making multiple request to Tenderly API
## Description
When you open a batch viewer transancion like:
<https://protocol-explorer.dev.gnosisdev.com/rinkeby/tx/0xaa5daa29e997cbf9b6307c707f13c76c3e2cb53ee41b8e85dc3e36e808cbbaf0/?tab=graph>

We can see how multiple requests are being made to get account addresses.

Selection_580

### Backgroud
This behavior has been inherited by introducing this feature #108

### PR DESCRIPTION
**Pull request by matextrem**
# Summary

Closes #121 

Updated Hook in order not to re-render it unnecessarily when there's no changes

# To Test

1. Go to a tx page in Batch Viewer view (`i.e: /rinkeby/tx/0xaa5daa29e997cbf9b6307c707f13c76c3e2cb53ee41b8e85dc3e36e808cbbaf0/?tab=graph`)
2.  Open network tab and search for tenderly requests
  * You will only see 2 requests (1 for getting trades and transfers and the other one for getting accounts).

### CODE REVIEW

**Review by henrypalacios — APPROVED** [review]
Nice catch!

File: src/hooks/useTxBatchTrades.tsx [hunk 1]
```diff
@@ -39,51 +39,48 @@ export function useTxBatchTrades(
...
   useEffect(() => {
-    if (!networkId || !ordersFoundInTx) {
+    if (!networkId || !orders?.length) {
       return
     }
 
-    _fetchTxTrades(networkId, txHash)
-  }, [_fetchTxTrades, networkId, ordersFoundInTx, txHash])
+    _fetchTxTrades(networkId, txHash, orders)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (ramirotw) [inline src/hooks/useTxBatchTrades.tsx:82]: why is this necessary?

Comment (henrypalacios) [reply src/hooks/useTxBatchTrades.tsx:82]: Good point, maybe it is not necessary, 
to avoid the rule breaking I can think of 2 things.
1. Continue to use `orders` and not execute `_fetchTrades` if it is **noLength/undfined**.
2. Or, pass it to `_fetchTrades` and stop it if there is **noLength/undfined** of `orders`.

Comment (matextrem) [reply src/hooks/useTxBatchTrades.tsx:82]: @ramirotw @henrypalacios The thing is not related to  it is undefined or not. The problem is the hook is executed twice once orders is populated (Since it is an array of objects, we would need to check for deep comparison in order not to execute it unnecessarily). So I'm using the orders length to prevent this.

Comment (ramirotw) [reply src/hooks/useTxBatchTrades.tsx:82]: I meant the eslint rule

Comment (henrypalacios) [reply src/hooks/useTxBatchTrades.tsx:82]: I understand @matextrem ! As you use `orders` as a dependency but it needs the **length** reference. 

To avoid disabling **eslint** we could pull the hook [`usePreviousLastValueData()`](https://github.com/cowprotocol/explorer/blob/develop/src/apps/explorer/components/SummaryCardsWidget/VolumeChart/VolumeChart.tsx#L213), to reuse it in cases like this.

Comment (matextrem) [reply src/hooks/useTxBatchTrades.tsx]: I finally found the problem! The problem was orders array was really changing. It seems like the first time we get the orders and the second time it is populated with `buyToken` and `sellToken` data so it caused the hook to re-render. Since we are not using those properties, I discarded them and use `previousValue` hook to look for any changes in order to improve the performance along with avoid unnecessary requests to tenderly API

**Review by elena-zh — APPROVED** [review]
LGTM

**Comment by alfetopito** [comment]
The behaviour described here is working nicely 👍 

However, I think I stumbled upon an edge case, which might or might not be related to the changes here.

To reproduce:

1. Load the PR url, on mainnet (the default) https://pr128--explorer.review.gnosisdev.com
2. Search for the tx 0xaa5daa29e997cbf9b6307c707f13c76c3e2cb53ee41b8e85dc3e36e808cbbaf0
3. Click on the graph tab
* The chart is not rendered
Screen Shot 2022-06-17 at 15 03 05

File: src/apps/explorer/components/SummaryCardsWidget/VolumeChart/VolumeChart.tsx [hunk 2]
```diff
@@ -69,13 +70,8 @@ function _formatAmount(amount: string): string {
...
 function usePreviousLastValueData<T>(value: T): T | undefined {
-  const ref = useRef<T>()
-
-  useEffect(() => {
-    ref.current = value
-  }, [value])
-
-  return ref.current
+  const previousValue = usePrevious(value)
+  return previousValue
```
Comment (alfetopito) [inline src/apps/explorer/components/SummaryCardsWidget/VolumeChart/VolumeChart.tsx:74]: Nit pick: What about returning the hook?

```suggestion
  return usePrevious(value)
```

Comment (alfetopito) [reply src/apps/explorer/components/SummaryCardsWidget/VolumeChart/VolumeChart.tsx:74]: Actually, unless there's anything special about the types, this hook is not needed and you can use `usePrevious` instead

**Comment by elena-zh** [comment]
> 2\. 0xaa5daa29e997cbf9b6307c707f13c76c3e2cb53ee41b8e85dc3e36e808cbbaf0

Hey @alfetopito , I can reproduce this issue only when I open the app for the 1st time. Then, all charts are loaded nicely.

**Comment by alfetopito** [comment]
> > To reproduce:
> > 
> > 1. Load the PR url, on mainnet (the default) https://pr128--explorer.review.gnosisdev.com
> > 2. Search for the tx 0xaa5daa29e997cbf9b6307c707f13c76c3e2cb53ee41b8e85dc3e36e808cbbaf0
> > 3. Click on the graph tab
> > 
> > 
> > * The chart is not rendered
> >   Screen Shot 2022-06-17 at 15 03 05
> 
> Hey @alfetopito , I can reproduce this issue only when I open the app for the 1st time. Then, all charts are loaded nicely.

Indeed, same as I observed.

**Comment by matextrem** [comment]
@alfetopito It seems it is not related to this PR. The tenderly API is returning 404 for that `txHash` in Rinkeby (IDK why in consequent times  it just works fine). I was not able to reproduce it locally.

Screen Shot 2022-06-17 at 21 44 25

**Comment by elena-zh** [comment]
I'm not able to reproduce this issue today

**Review by alfetopito — APPROVED** [review]

File: src/hooks/useTxBatchTrades.tsx [hunk 3]
```diff
@@ -37,53 +38,49 @@ export function useTxBatchTrades(
   const [error, setError] = useState('')
   const [txBatchTrades, setTxBatchTrades] = useState<TxBatchTrades>({ trades: [], transfers: [] })
   const [accounts, setAccounts] = useState<Accounts>()
+  const txOrders = usePrevious(JSON.stringify(orders?.map((o) => ({ owner: o.owner, kind: o.kind }))))
```
Comment (ramirotw) [inline src/hooks/useTxBatchTrades.tsx:41]: we should write some comment here to explain this tricky solution. Is not that obvious why the usePrevious hook and the stringifying are needed

**Review by ramirotw — CHANGES_REQUESTED** [review]

**Review by ramirotw — APPROVED** [review]
